src/uct/ugni/udt/ugni_udt_iface.c

```diff
@@ -185,24 +201,108 @@ static ucs_status_t uct_ugni_udt_iface_query(uct_iface_h tl_iface, uct_iface_att
+                datagram = iface->desc_any;
+                status = processs_datagram(iface, datagram);
+                if(UCS_OK != status) {
+                    UCT_TL_IFACE_GET_TX_DESC(&iface->super.super, &iface->free_desc,
+                                             iface->desc_any, iface->desc_any=NULL);
+                    user_desc = uct_ugni_udt_get_user_desc(datagram, iface);
+                    uct_recv_desc_iface(user_desc) = &iface->super.super.super;
+                }
+                pthread_mutex_lock(&uct_ugni_global_lock);
+                status = uct_ugni_udt_ep_any_post(iface);
+                pthread_mutex_unlock(&uct_ugni_global_lock);
+                if (UCS_OK != status) {
+                    /* We can't continue if we can't post the first receive */
+                    ucs_error("Failed to post wildcard request");
+                    return;
+                }
+            }
+        }
+        pthread_mutex_lock(&uct_ugni_global_lock);
+        ugni_rc = GNI_PostDataProbeById(iface->super.nic_handle, &id);
+        pthread_mutex_unlock(&uct_ugni_global_lock);
+    }
+
+    ucs_async_pipe_drain(&iface->event_pipe);
+    pthread_mutex_lock(&iface->device_lock);
+    iface->events_ready = 0;
+    pthread_mutex_unlock(&iface->device_lock);
+    ucs_trace("Signaling device thread to resume monitoring");
+    pthread_cond_signal(&iface->device_condition);
+
+}
+
+static void uct_ugni_udt_clean_wildcard(uct_ugni_udt_iface_t *iface)
 {
     gni_return_t ugni_rc;
+    uint32_t rem_addr, rem_id;
+    gni_post_state_t post_state;
+
+    pthread_mutex_lock(&uct_ugni_global_lock);
+    ugni_rc = GNI_EpPostDataCancelById(iface->ep_any, UCT_UGNI_UDT_ANY);
+    pthread_mutex_unlock(&uct_ugni_global_lock);
+    if (GNI_RC_SUCCESS != ugni_rc) {
+        ucs_error("GNI_EpPostDataCancel failed, Error status: %s %d\n",
+                  gni_err_str[ugni_rc], ugni_rc);
+    } else {
+        pthread_mutex_lock(&uct_ugni_global_lock);
+        ugni_rc = GNI_EpPostDataTestById(iface->ep_any, UCT_UGNI_UDT_ANY,
+                                         &post_state, &rem_addr, &rem_id);
+        pthread_mutex_unlock(&uct_ugni_global_lock);
+        if (GNI_RC_SUCCESS != ugni_rc && GNI_RC_NO_MATCH != ugni_rc) {
+            ucs_error("GNI_EpPostDataTestById failed, Error status: %s %d\n",
+                      gni_err_str[ugni_rc], ugni_rc);
+        }
+    }
+}
+
+static UCS_CLASS_CLEANUP_FUNC(uct_ugni_udt_iface_t)
+{
+    void *dummy;
 
-    uct_worker_progress_unregister(self->super.super.worker,
-                                   uct_ugni_udt_progress, self);
     if (!self->super.activated) {
         /* We done with release */
         return;
     }
 
-    ugni_rc = GNI_EpPostDataCancel(self->ep_any);
-    if (GNI_RC_SUCCESS != ugni_rc) {
-        ucs_debug("GNI_EpPostDataCancel failed, Error status: %s %d",
-                  gni_err_str[ugni_rc], ugni_rc);
-        return;
-    }
+    pthread_cancel(self->event_thread);
```

Comment:
good catch - it is not. You have to signal to the event thread that it is ok to shutdown.